### PR TITLE
feat(pipelines): add static PipelineBase.isPipeline method

### PR DESCRIPTION
**Pull request by Rabadash8820**
This change adds a new `isPipeline` method to `PipelineBase`. This method works the same way as `Stack.isStack` (checking if a specific `Symbol` property is defined on the provided object), and serves the same purpose: to check if a provided object extends `PipelineBase` (e.g., `CodePipeline` would return true, while `Stack` or `s3.Bucket` would return false).

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### ~Adding new Unconventional Dependencies:~

~* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)~

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by TheRealAmazonKendra — DISMISSED** [review]
Thank you for your contribution! I've exempted the requirements for a README update and integ-tests, but this will need unit tests. We really should have had unit tests for pipeline-base, but since we don't, feel free to create a new test file for it. 

Everything else looks good.

**Comment by Rabadash8820** [comment]
@TheRealAmazonKendra Sorry for the delay on this. I just added a unit test for the new `isPipeline` method, based on the unit test for `Stack.isStack`. As you mentioned above, there was no existing file for `PipelineBase` tests, so I added one at `packages/@aws-cdk/pipelines/test/main/pipeline-base.test.ts`. This mirrors the folder hierarchy of the `PipelineBase.ts` file itself. Note that, aside from the automated checks on this PR, I have _not_ actually run this new test myself (setting up builds/tests in VS Code was taking more effort than I cared to invest), so you or whoever is reviewing the code may want run it locally yourself.

**Comment by mrgrain** [comment]
@mergifyio update

**Comment by mrgrain** [comment]
Hey @Rabadash8820 change looks good but could you please make sure your code is using the correct styling. You can run the linter locally to fix or the PR build tells you any violations. Thanks.

**Comment by TheRealAmazonKendra** [comment]
> @TheRealAmazonKendra Sorry for the delay on this. I just added a unit test for the new `isPipeline` method, based on the unit test for `Stack.isStack`. As you mentioned above, there was no existing file for `PipelineBase` tests, so I added one at `packages/@aws-cdk/pipelines/test/main/pipeline-base.test.ts`. This mirrors the folder hierarchy of the `PipelineBase.ts` file itself. Note that, aside from the automated checks on this PR, I have _not_ actually run this new test myself (setting up builds/tests in VS Code was taking more effort than I cared to invest), so you or whoever is reviewing the code may want run it locally yourself.

I'm sorry, but no. We will not accept changes from contributors do not care to invest the effort to test their changes.

**Comment by Rabadash8820** [comment]
@TheRealAmazonKendra Wooooo that escalate quickly! I'm sorry, is the effort that I've put in to use your system, identify a potential improvement, clone your codebase and familiarize myself with it enough to implement the change _for you_, and then write a unit test to validate that the change works, all on my own time, not enough? Forgive me for being a little intimidated by all the build errors and CLI tools that this project uses (Lerna, in particular is still very new to me). I just assumed that, in the interest of closing this issue, it would be simpler for those that are _familiar_ with this codebase to actually run the test, one that I'm very confident will pass because I _took the effort_ to emulate your existing tests.

I understand that this is a small change, so I won't pretend that merging it isn't the most important thing in the world, but good lord, give me another chance to run the test myself then. Insulting and belittling your contributors is not the way to build a healthy open source community. And what's more, _I work at Amazon_ (look me up on PhoneTool), so yeah I'm very insulted at being chastised by another developer at my own company, and frankly embarrassed that this is now the open-source conduct that _I and my team_ will be associated with.

**TL;DR** Please re-open this issue so that I can address @mrgrain's styling concerns and make another attempt at running the new unit test.

**Comment by Rabadash8820** [comment]
@mrgrain Which PR check shows the styling violations? I don't have access to the AWS account linked from the `AWS CodeBuild us-east-1` check, and the `PR Linter` check shows everything as passing. 🤔

**Comment by mrgrain** [comment]
@Rabadash8820 I've been there. The build posts a comment with link to the public build logs.
https://github.com/aws/aws-cdk/pull/21075#issuecomment-1209215774

**Comment by TheRealAmazonKendra** [comment]
As @mrgrain notes, we have multiple linters. The PR linter only informs you of what is missing from the PR, i.e. unit tests, integration tests, or a README update. We also run eslint, pkglint, and awslint in the build. 
- [eslint](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#eslint)
- [pkglint](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#pkglint)
- [awslint](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#awslint)

If you look at the message posted by aws-cdk-automation, you can see the output of the build (pasted below for your convenience):

> AWS CodeBuild CI Report
CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
Commit ID: https://github.com/aws/aws-cdk/commit/ab777921eae4184865439bedba3b746549405f9a
Result: FAILED
[Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7365652a-8f88-4990-b58e-6509662711f0%2Fbuild.log) (available for 30 days)
Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/):

The build logs link will take our straight to the build output so you can see any failure messages, errors, etc. We cannot accept code changes that do not pass the build.

With ~65 open PRs and ~1700 open issues, we do not have the capacity to fix tests and code on behalf of contributors. We appreciate the work of everyone and we are happy to answer any questions about failures/error/unclear documentation.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Putting this back into changes requested so we can see more clearly when updates have been made.

**Comment by Rabadash8820** [comment]
Thanks for the pointers, @mrgrain and @TheRealAmazonKendra. I've fixed those styling issues and updated the unit test so that it actually passes (see latest commit message for more info). It took a little while cause the `yarn build` and `yarn test` commands were running really slow, which made it hard to iterate on the unit test and its test `PipelineBase` implementation. I think this might be because I was running the commands in WSL but the actual source files were cloned in my Windows filesystem. I haven't tried cloning inside WSL yet, but I bet that that would improve performance of the `yarn` commands.
